flang/docs/ParallelFortranRuntime.md

```diff
@@ -0,0 +1,1987 @@
+  after [`prif_change_team`](#prif_change_team), or to pass to a coarray dummy
+  argument (especially in the case that the cobounds are different)
+    
+```
+subroutine prif_alias_create( &
+    source_handle, alias_lcobounds, alias_ucobounds, alias_handle)
+  type(prif_coarray_handle), intent(in) :: source_handle
+  integer(c_intmax_t), intent(in) :: alias_lcobounds(:)
+  integer(c_intmax_t), intent(in) :: alias_ucobounds(:)
+  type(prif_coarray_handle), intent(out) :: alias_handle
+end subroutine
+```
+**Further argument descriptions**:
+
+* **`source_handle`**: a handle (which may itself be an alias) to the existing
+  coarray for which an alias is to be created
+* **`alias_lcobounds` and `alias_ucobounds`**: the cobounds to be used for
+  the new alias. Both arguments must have the same size, but it need not
+  match the corank associated with `source_handle`
+* **`alias_handle`**: a new alias to the existing coarray
+
+### `prif_alias_destroy`
+
+**Description**: Delete an alias to a coarray. Does not deallocate the original coarray.
+    
+```
+subroutine prif_alias_destroy(alias_handle)
+  type(prif_coarray_handle), intent(in) :: alias_handle
+end subroutine
+```
+**Further argument descriptions**:
+
+* **`alias_handle`**: the alias to be destroyed
+
+### `MOVE_ALLOC`
+
+This is not provided by PRIF because it depends on unspecified details
+of the compiler's `allocatable` attribute. It is the compiler's responsibility
+to implement `MOVE_ALLOC` using PRIF-provided operations. For example, according
+to the Fortran standard, `MOVE_ALLOC` with coarray arguments is an image control statement that
+requires synchronization, so the compiler should likely insert call(s) to
+`prif_sync_all` as part of the implementation.
+
+| **CLIENT NOTE**: |
+| ---------------- |
+| It is envisioned that the use of `prif_set_context_data` and `prif_get_context_data` will allow for an efficient implementation of `MOVE_ALLOC` that maintains tracking of allocation status |
+
+## Coarray Queries
+
+### `prif_set_context_data`
+
+**Description**: This procedure stores a `c_ptr` associated with a coarray
+  for future retrieval. A typical usage would be to store a reference
+  to the actual variable whose allocation status must be changed in the case
+  that the coarray is deallocated.
+    
+```
+subroutine prif_set_context_data(coarray_handle, context_data)
+  type(prif_coarray_handle), intent(in) :: coarray_handle
+  type(c_ptr), intent(in) :: context_data
+end subroutine
+```
+
+### `prif_get_context_data`
+
+**Description**: This procedure returns the `c_ptr` provided in the most
+  recent call to [`prif_set_context_data`](#prif_set_context_data) with the
+  same coarray (possibly via an alias coarray handle).
+    
+```
+subroutine prif_get_context_data(coarray_handle, context_data)
+  type(prif_coarray_handle), intent(in) :: coarray_handle
+  type(c_ptr), intent(out) :: context_data
+end subroutine
+```
+
+### `prif_base_pointer`
+
```

Comment:
Quoted from @jeffhammond's response:

> `prif_base_pointer` does not describe the failure mode when memory is not accessible via load-store at the specified image. This function seems equivalent to shmem_ptr, which is documented to return NULL when remote memory is not accessible via pointers.
Assuming I infer correctly that this function plus the raw operations are designed to optimize for shared-memory communication via load-store, there needs to be a high-level semantic overview of the purpose of these functions and their difference from the non-raw operations (which need to include strided variants, as noted above).
If I do not infer correctly the use case, then I disagree with the design.
Side note: I have implemented OpenSHMEM on top of MPI with the general case and specialization for the shared-memory path (https://github.com/jeffhammond/oshmpi/blob/master/src/shmem-internals.c). It is going to be important to support both use cases because we want coarrays to be competitive with shared-memory parallelism (e.g. OpenMP) when programs run within a shared-memory domain.

Reply:
**LBL response:** `prif_base_pointer` is not equivalent or even similar to `shmem_ptr`, so this is not the right intuition. `prif_base_pointer` gives you a pointer that is usable by the target process, not by the caller. This has nothing to do with potential shared memory bypass and is not an optional feature. 
We acknowledge that PRIF interface-level support for explicit shared-memory bypass optimizations by the compiler may eventually be important. The current draft does not yet include any features to explicitly support shared-memory bypass in the caller, however we believe that such features can be added later as extensions in future PRIF revisions, without disrupting the current PRIF design. Based on your feedback, we will add this to the future work section.
The core issue here is that Fortran allows remote access (via put/get, atomics, and notify/events/locks) to memory that was not allocated during any coarray allocation. One specific example of this is a coarray of derived type with an ALLOCATABLE or POINTER component. That component is allocated non-collectively by the image with affinity, and the allocation of the target memory backing that component can happen long after the coarray itself is allocated, or even before in the case of a POINTER component. As such, there is no guaranteed “binding” between the storage backing the coarray and the storage backing these indirected components (in particular, they need not reside at a fixed offset from the base of the coarray storage) yet they are nevertheless remotely accessible by Fortran semantics.
This language semantic imposes a requirement on compliant runtime implementations, and in turn on the PRIF specification to expose capabilities fulfilling this requirement. PRIF currently deals with this situation by exposing the ability to perform “raw”-style operations, where the remote target is identified by a raw pointer (represented as a c_intptr_t) that is the target image’s virtual memory address of the target storage. This is the motivation for the `prif_*{put,get}*_raw` procedures, and currently the PRIF subsystems for atomics, locks, events and notify all expose only “raw”-style interfaces; this is because a “raw”-style interface is general enough to encompass any use case (i.e. whether the target location is stored directly in the coarray or indirected from the coarray storage through one or more ALLOCATABLE or POINTER components).
`prif_base_pointer` is currently the “starting point” by which the compiler’s generated code can traverse through (in general a series of) indirect coarray components on a remote image. In the case of access to an indirect coarray component, the generated code would likely begin with a call to `prif_base_pointer` followed by a `prif_get_raw` call to retrieve the bytes corresponding to the first pointer in the chain of indirection. PRIF is carefully written to avoid assuming anything about the compiler’s representation of POINTER or ALLOCATABLE components, which is part of why `prif_get_raw` allows the caller to fetch an arbitrary contiguous block of memory from the target’s shared heap. After retrieving the bytes corresponding to the first pointer component, the caller would interpret those bytes in a compiler-specific manner to determine the next raw pointer in the chain of indirection, and pass it in another call to a “raw” style interface. The chain ends when a raw pointer to the ultimate storage location on the target image has been determined, and the caller finally invokes a raw-style procedure to perform the actual operation (put, get, atomic, lock, etc). 
If this approach sounds expensive, that’s because it probably is. One-sided indirection through POINTER or ALLOCATABLE components in a coarray fundamentally requires either a series of initiator-chained multiple round-trips as described here, or a more general active-message like facility (which would represent an even further departure from anything in Fortran). Such indirection hopefully never appears as a performance-critical operation in well-written parallel Fortran programs, but the language requires they must work, so we need to specify the means for a correct/compliant implementation of the indirected access case.

Reply:
```fortran
program put_example
   implicit none


   type :: t
       integer, allocatable :: val
   end type


   type(t) :: obj[*]


   call do_alloc(obj%val)
   sync all
   if (this_image() == 2) obj[1]%val = 42
   sync all
   if (this_image() == 1) print *, obj%val
contains
   subroutine do_alloc(v)
       integer, allocatable, intent(out) :: v
       allocate(v)
   end subroutine
end program
```

```fortran
program atomic_example
   use iso_fortran_env, only: atomic_int_kind
   implicit none


   type :: t
       integer(atomic_int_kind), allocatable :: val
   end type


   type(t) :: obj[*]


   call do_alloc(obj%val)
   sync all
   call atomic_add(obj[1]%val, 1)
   sync all
   if (this_image() == 1) print *, obj%val
contains
   subroutine do_alloc(v)
       integer(atomic_int_kind), allocatable, intent(out) :: v
       allocate(v)
   end subroutine
end program
```

Reply:
>PRIF is carefully written to avoid assuming anything about the compiler’s representation of POINTER or ALLOCATABLE components, which is part of why prif_get_raw allows the caller to fetch an arbitrary contiguous block of memory from the target’s shared heap. 

I don't understand this assertion.  `prif_base_pointer` returns an `c_intptr_t`, which is not sufficient to access remote memory on systems that require memory registration.  If you want a portable implementation, you need an opaque handle that can include the memory registration key.

> After retrieving the bytes corresponding to the first pointer component, the caller would interpret those bytes in a compiler-specific manner to determine the next raw pointer in the chain of indirection, and pass it in another call to a “raw” style interface.

As noted above, this is not a portable assumption.  Your design effectively assumes GASNet.

Reply:
The interfaces for direct accesses added in v0.4 resolve this concern.